### PR TITLE
CI: set concurrency

### PR DESCRIPTION
**Pull request by Borda**
Found this use-case for GH action and Tox - https://github.com/tox-dev/tox-gh; what do you think?

### CODE REVIEW

**Comment by hugovk** [comment]
I don't really understand what tox-gh brings, beyond duplicating the Python version numbers in test.yml and in tox.ini

Please could you elaborate on the benefits?

**Comment by Borda** [comment]
> Please could you elaborate on the benefits?

You are probably right, just found it easier to read/follow...

**Comment by JulienPalard** [comment]
I'm not a fan of bringing yet another dependency here. If it was me alone I would just have a single github build (with multiple actions/setup-python in it, yes it works) running a single tox instance: trying to keep it simple, a bit like in https://github.com/sphinx-contrib/sphinx-lint/commit/45af71c1fc6cd56a8e5d6e064e36f5e69edf211a.

**Comment by Borda** [comment]
> it was me alone I would just have a single github build (with multiple actions/setup-python in it, yes it works) running a single tox instance: trying to keep it simple

Yes, agree but since the Tox is here, it is better to check it and the simplest was to do so is use it... So to say reduce dead code =) 

> a bit like in [45af71c](https://github.com/sphinx-contrib/sphinx-lint/commit/45af71c1fc6cd56a8e5d6e064e36f5e69edf211a).

Well, this one is also strange with redundant python setups...

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -1,52 +1,58 @@
...
-name: Tests
+name: Tests with Tox
 
 on: [push, pull_request, workflow_dispatch]
 
 env:
   FORCE_COLOR: 1
 
+concurrency:
+  group: check-tests-tox-${{ github.ref }}
```
Comment (hugovk) [inline .github/workflows/tests.yml:9]:
```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -1,52 +1,58 @@
...
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
```
Comment (hugovk) [inline .github/workflows/tests.yml:10]: Also cancel repeat builds on `main`?
```suggestion
  cancel-in-progress: true
```

**Review by hugovk — CHANGES_REQUESTED** [review]
Anyway, back to this PR.

I think the `concurrency:` changes are useful, and let's revert the others.

**Comment by Borda** [comment]
> Anyway, back to this PR.

OK will update PR and it's name accordingly

File: .github/workflows/tests.yml (hunk 2 above)
Comment (Borda) [reply .github/workflows/tests.yml:10]: I would keep it running on main branch always, so how about cancel of not push to main?

Comment (hugovk) [reply .github/workflows/tests.yml:10]: No strong preference, this isn't a very busy repo so that sounds fine.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -1,12 +1,15 @@
----
-
-name: Tests
+name: Tests with Tox
```
Comment (hugovk) [inline .github/workflows/tests.yml:1]:
```suggestion
---
name: Tests
```

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -34,16 +37,13 @@ jobs:
 
       - name: Install dependencies
         run: |
-          python -m pip install -U pip
-          python -m pip install -U wheel
-          python -m pip install -U tox
+          pip install -U wheel tox
```
Comment (hugovk) [inline .github/workflows/tests.yml:40]: I think we can skip wheel, but let's make sure pip is latest.
```suggestion
          pip install -U pip
          pip install -U tox
```

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -34,16 +37,13 @@ jobs:
...
 
       - name: Download more tests from friend projects
         if: matrix.python-version == '3.13' && matrix.os == 'ubuntu-latest'
         run: sh download-more-tests.sh
       - name: Tox tests
-        run: |
-          tox -e py
+        run: tox -e py
```
Comment (hugovk) [inline .github/workflows/tests.yml:46]:
```suggestion
        run: |
          tox -e py
```

File: .github/workflows/lint.yml [hunk 6]
```diff
@@ -8,6 +8,11 @@ env:
 permissions:
   contents: read
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  # cancel all except push to main branch for have always full results
```
Comment (hugovk) [inline .github/workflows/lint.yml:13]:
```suggestion
  # cancel all except push to main branch to have always full results
```

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -1,12 +1,15 @@
...
+name: Tests with Tox
 
 on: [push, pull_request, workflow_dispatch]
 
 env:
   FORCE_COLOR: 1
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  # cancel all except push to main branch for have always full results
```
Comment (hugovk) [inline .github/workflows/tests.yml:10]:
```suggestion
  # cancel all except push to main branch to have always full results
```

**Review by hugovk — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 4 above)
Comment (Borda) [reply .github/workflows/tests.yml:40]: Why reinstall pip?

**Review by hugovk — APPROVED** [review]
Thanks!

**Comment by Borda** [comment]
@hugovk so fine to land? :chipmunk:

**Comment by Borda** [comment]
@hugovk is it fine now, or shall I update something else? :flamingo:

**Pull request merged**